File: opc2aas/OPC2AAS/DatabridgeConfig.cs
using Newtonsoft.Json;

namespace OPC2AAS
{
    /// <summary>
    /// Class for the Databridge config creation
    /// </summary>
    public class DatabridgeConfig
    {
        /// <summary>
        /// Method that reates new Databridge config files and clears old ones.
        /// </summary>
        public void CreateOrClearConfigFiles()
        {
            // check if config files exist in Content Folder
            // Files: aasserver.json, jsonataExtractValue.json, jsonatatransformer.json, jsonjacksontransformer.json, opcuaconsumer.json, routes.json, timerconsumer.json
            string[] files = { "aasserver.json", "jsonataExtractValue.json", "jsonatatransformer.json", "jsonjacksontransformer.json", "opcuaconsumer.json", "routes.json", "timerconsumer.json" };
            foreach (string file in files)
            {
                string filePath = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "Content", file);
                // create the config file if it does not exist
                if (!File.Exists(filePath))
                {
                    using (FileStream fs = File.Create(filePath)) { }
                    
                }
                // write the respective content to the config file (empty array or default values)
                if (file == "jsonataExtractValue.json")
                {
                    File.WriteAllText(filePath, "value.value");
                }
                else if (file == "jsonjacksontransformer.json")
                {
                    File.WriteAllText(filePath, "[\r\n  {\r\n    \"uniqueId\": \"dataValueToJson\",\r\n    \"operation\": \"marshal\",\r\n    \"jacksonModules\": \"com.fasterxml.jackson.datatype.jsr310.JavaTimeModule\"\r\n  }\r\n]");
                }
                else if (file == "jsonatatransformer.json")
                {
                    File.WriteAllText(filePath, "[\r\n  {\r\n    \"uniqueId\": \"jsonataExtractValue\",\r\n    \"queryPath\": \"jsonataExtractValue.json\",\r\n    \"inputType\": \"JsonString\",\r\n    \"outputType\": \"JsonString\"\r\n  }\r\n]");
                }
                else if (file == "timerconsumer.json")
                {
                    File.WriteAllText(filePath, "[\r\n  {\r\n    \"uniqueId\": \"timer1\",\r\n    \"fixedRate\": true,\r\n    \"delay\": 0,\r\n    \"period\": 1000\r\n  }\r\n]");
                }
                else
                {
                    File.WriteAllText(filePath, "[]");
                }
            }
        }
        /// <summary>
        /// Writes the OPC UA consumer Databridge config file.
        /// </summary>
        /// <param name="OPCAddress">The OPC address.</param>
        /// <param name="currentNode">The current node.</param>
        /// <param name="submodelName">Name of the parent Submodel.</param>
        /// <param name="idShortPath">Path for the current SubmodelElement (layer).</param>
        public void WriteOpcUaConsumer(string OPCAddress, OPCUtil.Node currentNode, string submodelName, string idShortPath)
        {
            string[] parts = OPCAddress.Split(':');
            string serverUrl = parts[0];
            // change localhost to host.docker.internal (needed when running in a Docker container)
            if (serverUrl == "localhost")
            {
                serverUrl = "host.docker.internal";
            }
            int serverPort = int.Parse(parts[1]);

            string filePath = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "Content", "opcuaconsumer.json");
            string jsonContent = File.ReadAllText(filePath);

            List<OpcEntry> entries = JsonConvert.DeserializeObject<List<OpcEntry>>(jsonContent);

            OpcEntry newEntry = new OpcEntry
            {
                uniqueId = "source://" + submodelName + "/" + idShortPath + currentNode.name,
                serverUrl = serverUrl,
                serverPort = serverPort,
                pathToService = "",
                nodeInformation = currentNode.nodeId,
                requestPublishingInterval = 50
            };
            entries.Add(newEntry);

            string newJsonContent = JsonConvert.SerializeObject(entries, Formatting.Indented);

            File.WriteAllText(filePath, newJsonContent);
        }
        /// <summary>
        /// Writes the routes Databridge config file.
        /// </summary>
        /// <param name="currentNode">The current node.</param>
        /// <param name="submodelName">Name of the parent Submodel.</param>
        /// <param name="idShortPath">Path for the current SubmodelElement (layer).</param>
        public void WriteRoutes(OPCUtil.Node currentNode, string submodelName, string idShortPath)
        {
            string filePath = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "Content", "routes.json");
            string jsonContent = File.ReadAllText(filePath);

            List<RouteEntry> entries = JsonConvert.DeserializeObject<List<RouteEntry>>(jsonContent);
            
            RouteEntry newEntry = new RouteEntry
            {
                datasource = "source://" + submodelName + "/" + idShortPath + currentNode.name,
                transformers = new string[] { "dataValueToJson", "jsonataExtractValue" },
                datasinks = new string[] { "sink://" + submodelName + "/" + idShortPath + currentNode.name },
                trigger = "timer",
                triggerData = new TriggerData
                {
                    timerName = "timer1"
                }
            };

            entries.Add(newEntry);

            string newJsonContent = JsonConvert.SerializeObject(entries, Formatting.Indented);

            File.WriteAllText(filePath, newJsonContent);
        }
        /// <summary>
        /// Writes the aas server.
        /// </summary>
        /// <param name="currentNode">The current node.</param>
        /// <param name="submodelName">Name of the parent Submodel.</param>
        /// <param name="AASName">Name of the to be generated AAS.</param>
        /// <param name="idShortPath">Path for the current SubmodelElement (layer).</param>
        public void WriteAASServer(OPCUtil.Node currentNode, string submodelName, string AASName, string idShortPath)
        {
            string filePath = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "Content", "aasserver.json");
            string jsonContent = File.ReadAllText(filePath);

            List<AasEntry> entries = JsonConvert.DeserializeObject<List<AasEntry>>(jsonContent);

            AasEntry newEntry = new AasEntry
            {
                uniqueId = "sink://" + submodelName + "/" + idShortPath + currentNode.name,
                submodelEndpoint = "http://host.docker.internal:4500/aasServer/shells/" + AASName + "/aas/submodels/" + submodelName + "/submodel",
                idShortPath = idShortPath + currentNode.name
            };

            entries.Add(newEntry);

            string newJsonContent = JsonConvert.SerializeObject(entries, Formatting.Indented);

            File.WriteAllText(filePath, newJsonContent);
        }
    }
    /// <summary>
    /// Class for the OPC UA consumer datastructure.
    /// </summary>
    class OpcEntry
    {
        public string uniqueId { get; set; }
        public string serverUrl { get; set; }
        public int serverPort { get; set; }
        public string pathToService { get; set; }
        public string nodeInformation { get; set; }
        public int requestPublishingInterval { get; set; }
    }
    /// <summary>
    /// Class for the routes datastructure.
    /// </summary>
    class RouteEntry
    {
        public string datasource { get; set; }
        public string[] transformers { get; set; }
        public string[] datasinks { get; set; }
        public string trigger { get; set; }
        public TriggerData triggerData { get; set; }
    }
    /// <summary>
    /// Class for the trigger datastructure.
    /// </summary>
    public class TriggerData
    {
        public string timerName { get; set; }
    }
    /// <summary>
    /// Class for the AAS server datastructure.
    /// </summary>
    public class AasEntry
    {
        public string uniqueId { get; set; }
        public string submodelEndpoint { get; set; }
        public string idShortPath { get; set; }
    }
}

File: opc2aas/opc2aas/src/main/java/org/eclipse/digitaltwin/basyx/opc2aas/DataBridgeConfig.java
package org.eclipse.digitaltwin.basyx.opc2aas;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;

import com.google.common.hash.Hashing;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.List;
import java.util.Base64;

public class DataBridgeConfig {

    private static final Logger logger = LoggerFactory.getLogger(AasBuilder.class);

    /**
     * A class representing an entry in the opcuaconsumer.json file.
     */
    public static class OpcUaConsumerEntry {
        public String uniqueId;
        public String serverUrl;
        public int serverPort;
        public String pathToService;
        public String nodeInformation;
        @JsonInclude(JsonInclude.Include.NON_EMPTY) // This field will be skipped if it is null or an empty string
        public String username;
        @JsonInclude(JsonInclude.Include.NON_EMPTY) // This field will be skipped if it is null or an empty string
        public String password;
        public int requestPublishInterval;

        public OpcUaConsumerEntry() {
            // No-arg constructor for Jackson
        }

        public OpcUaConsumerEntry(String uniqueId, String serverUrl, int serverPort, String pathToService, String nodeInformation, String username, String password, int requestPublishInterval) {
            this.uniqueId = uniqueId;
            this.serverUrl = serverUrl;
            this.serverPort = serverPort;
            this.pathToService = pathToService;
            this.nodeInformation = nodeInformation;
            this.username = username;
            this.password = password;
            this.requestPublishInterval = requestPublishInterval;
        }
    }

    /**
     * A class representing an entry in the routes.json file.
     */
    public static class RouteEntry {
        public String datasource;
        public String[] transformers;
        public String[] datasinks;
        public String trigger;

        public RouteEntry() {
            // No-arg constructor for Jackson
        }

        public RouteEntry(String datasource, String[] transformers, String[] datasinks, String trigger) {
            this.datasource = datasource;
            this.transformers = transformers;
            this.datasinks = datasinks;
            this.trigger = trigger;
        }
    }

    /**
     * A class representing an entry in the aasserver.json file.
     */
    public static class AasServerEntry {
        public String uniqueId;
        public String submodelEndpoint;
        public String idShortPath;
        public String api;

        public AasServerEntry() {
            // No-arg constructor for Jackson
        }

        public AasServerEntry(String uniqueId, String submodelEndpoint, String idShortPath, String api) {
            this.uniqueId = uniqueId;
            this.submodelEndpoint = submodelEndpoint;
            this.idShortPath = idShortPath;
            this.api = api;
        }
    }

    /**
     * Creates the initial configuration files for the DataBridge.
     */
    public static void createConfigFiles() {

        String[] fileNames = {
                "aasserver.json",
                "jsonataExtractValue.json",
                "jsonatatransformer.json",
                "jsonjacksontransformer.json",
                "opcuaconsumer.json",
                "routes.json"
        };

        for (String fileName : fileNames) {
            try {
                File file = new File(fileName);
                FileWriter writer = new FileWriter(file, false); // false to overwrite

                switch (fileName) {
                    case "jsonataExtractValue.json":
                        writer.write("value.value");
                        break;

                    case "jsonjacksontransformer.json":
                        writer.write("[\r\n  {\r\n    \"uniqueId\": \"dataValueToJson\",\r\n    \"operation\": \"marshal\",\r\n    \"jacksonModules\": \"com.fasterxml.jackson.datatype.jsr310.JavaTimeModule\"\r\n  }\r\n]");
                        break;

                    case "jsonatatransformer.json":
                        writer.write("[\r\n  {\r\n    \"uniqueId\": \"jsonataExtractValue\",\r\n    \"queryPath\": \"jsonataExtractValue.json\",\r\n    \"inputType\": \"JsonString\",\r\n    \"outputType\": \"JsonString\"\r\n  }\r\n]");
                        break;

                    default:
                        writer.write("[]");
                        break;
                }

                writer.close();
            } catch (IOException e) {
                logger.error("An error occurred while creating the DataBridge configuration files: ", e);
            }
        }
    }

    /**
     * Writes an entry to the opcuaconsumer.json file.
     *
     * @param endpointUrl The endpoint URL of the OPC UA server.
     * @param username The username to use for the connection.
     * @param password The password to use for the connection.
     * @param node The NodeInfo object of the node to create an entry for.
     * @param submodelIdShort The idShort of the submodel.
     * @param idShortPath The path to the submodelElement.
     */
    public static void writeOpcUaConsumerEntry(String endpointUrl, String username, String password, NodeInfo node, String submodelIdShort, String idShortPath) {
        try {
            ObjectMapper objectMapper = new ObjectMapper()
                    .setSerializationInclusion(JsonInclude.Include.NON_NULL); // Ignore null and empty fields globally;

            File file = new File("opcuaconsumer.json");
            // Read existing content
            List<OpcUaConsumerEntry> existingEntries = new ArrayList<>();
            if (file.exists()) {
                String currentContent = Files.readString(file.toPath());
                existingEntries = objectMapper.readValue(currentContent, new TypeReference<List<OpcUaConsumerEntry>>() {});
            }

            // Create a new entry
            String idShort = node.displayName.getText();
            String hashedSourcePath = Hashing.sha256().hashString("source://" + submodelIdShort + "/" + idShortPath + idShort + node.nodeId.toParseableString() + endpointUrl, StandardCharsets.UTF_8).toString(); // hash the source path

            OpcUaConsumerEntry newEntry = new OpcUaConsumerEntry(
                    hashedSourcePath,
                    endpointUrl.split(":")[1].substring(2).equals("localhost") ? "host.docker.internal" : endpointUrl.split(":")[1].substring(2),
                    Integer.parseInt(endpointUrl.split(":")[2].split("/")[0]),
                    "",
                    node.nodeId.toParseableString(),
                    username,
                    password,
                    1000
            );

            // Append the new entry
            existingEntries.add(newEntry);

            // Write the updated content back to the file
            FileWriter writer = new FileWriter(file, false); // false to overwrite
            writer.write(objectMapper.writerWithDefaultPrettyPrinter().writeValueAsString(existingEntries));
            writer.close();
        } catch (IOException e) {
            logger.error("An error occurred while writing to the opcuaconsumer.json file: ", e);
        }
    }

    /**
     * Writes an entry to the routes.json file.
     *
     * @param node The NodeInfo object of the node to create an entry for.
     * @param submodelIdShort The idShort of the submodel.
     * @param idShortPath The path to the submodelElement.
     */
    public static void writeRouteEntry(NodeInfo node, String submodelIdShort, String idShortPath, String endpointUrl) {
        try {
            ObjectMapper objectMapper = new ObjectMapper();

            // Read existing content
            File file = new File("routes.json");
            List<RouteEntry> existingEntries = new ArrayList<>();
            if (file.exists()) {
                String currentContent = Files.readString(file.toPath());
                existingEntries = objectMapper.readValue(currentContent, new TypeReference<List<RouteEntry>>() {});
            }

            // Create a new entry
            String idShort = node.displayName.getText();
            String hashedSourcePath = Hashing.sha256().hashString("source://" + submodelIdShort + "/" + idShortPath + idShort + node.nodeId.toParseableString() + endpointUrl, StandardCharsets.UTF_8).toString(); // hash the source path
            String hashedSinkPath = Hashing.sha256().hashString("sink://" + submodelIdShort + "/" + idShortPath + idShort  +  endpointUrl, StandardCharsets.UTF_8).toString(); // hash the sink path
            String[] transformers = {"dataValueToJson", "jsonataExtractValue"};
            String[] datasinks = {hashedSinkPath};

            RouteEntry newEntry = new RouteEntry(
                    hashedSourcePath,
                    transformers,
                    datasinks,
                    "event"
            );

            // Append the new entry
            existingEntries.add(newEntry);

            // Write the updated content back to the file
            FileWriter writer = new FileWriter(file, false); // false to overwrite
            writer.write(objectMapper.writerWithDefaultPrettyPrinter().writeValueAsString(existingEntries));
            writer.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    /**
     * Writes an entry to the aasserver.json file.
     *
     * @param node The NodeInfo object of the node to create an entry for.
     * @param submodelIdShort The idShort of the submodel.
     * @param submodelIdentifier The identifier of the submodel.
     * @param idShortPath The path to the submodelElement.
     * @param submodelRepositoryUrl The URL of the AAS Repository.
     */
    public static void writeAasServerEntry(NodeInfo node, String submodelIdShort, String submodelIdentifier, String idShortPath, String submodelRepositoryUrl, String endpointUrl) {
        try {
            ObjectMapper objectMapper = new ObjectMapper();

            // Read existing content
            File file = new File("aasserver.json");
            List<AasServerEntry> existingEntries = new ArrayList<>();
            if (file.exists()) {
                String currentContent = Files.readString(file.toPath());
                existingEntries = objectMapper.readValue(currentContent, new TypeReference<List<AasServerEntry>>() {});
            }

            // Create a new entry
            String idShort = node.displayName.getText();
            String hashedSinkPath = Hashing.sha256().hashString("sink://" + submodelIdShort + "/" + idShortPath + idShort  +  endpointUrl, StandardCharsets.UTF_8).toString(); // hash the sink path
            String submodelEndpoint = submodelRepositoryUrl + "/" + base64Encode(submodelIdentifier);
            String api = "DotAAS-V3";

            AasServerEntry newEntry = new AasServerEntry(
                    hashedSinkPath,
                    submodelEndpoint,
                    idShortPath + idShort,
                    api
            );

            // Append the new entry
            existingEntries.add(newEntry);

            // Write the updated content back to the file
            FileWriter writer = new FileWriter(file, false); // false to overwrite
            writer.write(objectMapper.writerWithDefaultPrettyPrinter().writeValueAsString(existingEntries));
            writer.close();
        } catch (IOException e) {
            logger.error("An error occurred while writing to the aasserver.json file: ", e);
        }
    }

    /**
     * Encodes a string to base64.
     *
     * @param value The string to encode.
     * @return The encoded string.
     */
    public static String base64Encode(String value) {
        byte[] bytes = value.getBytes(StandardCharsets.UTF_8);
        return Base64.getEncoder().encodeToString(bytes);
    }
}
